### PR TITLE
Revert "util: Simplify non-copy string-[]byte conversion"

### PR DESCRIPTION
**Pull request by suzaku**
This reverts commit e954a5a12239a1b8dab01d68f89f1270b08762a5.

### What problem does this PR solve? 

Commit `e954a5a12239a1b8dab01d68f89f1270b08762a5` causes a reproducible performance
regression for the case `insert_batch`.

The original implementation would return directly if the input is empty,
so we try to fix this problem by adding the shortcut for empty inputs in
#11828.

But it turns out the regression is still there.

### What is changed and how it works?

Revert the change.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### CODE REVIEW

**Comment by mahjonp** [comment]
> could we add a benchmark test for the case you mentioned?

We have a `batch insert` benchmark, and you can review it on perf.pingcap.net

File: util/hack/hack.go [hunk 1]
```diff
@@ -24,8 +24,15 @@ type MutableString string
 // String converts slice to MutableString without copy.
 // The MutableString can be converts to string without copy.
 // Use it at your own risk.
-func String(b []byte) MutableString {
-	return *(*MutableString)(unsafe.Pointer(&b))
+func String(b []byte) (s MutableString) {
```
Comment (SunRunAway) [inline util/hack/hack.go]: Can you provide a [benchmark](https://golang.org/pkg/testing/#hdr-Benchmarks) for this function and provide comparable results before and after.

**Comment by mahjonp** [comment]
@suzaku maybe we should update branch and benchmark again

**Comment by zz-jason** [comment]
@suzaku Can we close this PR? Seems it's not help with the performance?

**Comment by suzaku** [comment]
@mahjonp According to the benchmark result, reverting the change doesn't seem to fix the performance regression?

**Comment by mahjonp** [comment]
@zz-jason sorry, I will bench with batch_insert workload later, please wait a moment.

**Comment by mahjonp** [comment]
```
Version:
tidb: f36fff303191a32d553659a09efa2da7664765a7(HEAD) 2019-08-30 12:48:23
tikv: d2e3b6af4226b17595e361da93ffcb48b89e7ec5(master) 2019-09-11 03:12:39
pd: e678a1f5c022fed729fb79397fe02b6c9f54ff4a(master) 2019-09-10 06:57:15

tidb_max_cpu: 21.63
tikv_max_cpu: 8.60
tidb_max_memory: 5160.36 MiB
tikv_max_memory: 9499.29 MiB

tps: 469.27 ± 1.4712% (std=3.81)
qps: 469.27 ± 1.4712% (std=3.81)
lat_avg: 545.37 ± 1.4601% (std=4.43)
lat_99th: 1170.73 ± 1.8096% (std=14.05)

vs

Version:
tidb: 0f2434660c509ec0b493d8c7ee0bbbf8e8549184(master) 2019-09-09 11:33:43
tikv: ff82aa9eba331585aec1c6cdf9e1584512bccb34(master) 2019-09-09 08:37:54
pd: c7c572a7dc9710aca57cb0bddf4d9bca6c4a111b(master) 2019-09-09 09:30:29

tidb_max_cpu: 22.53
tikv_max_cpu: 7.59
tidb_max_memory: 5373.36 MiB
tikv_max_memory: 7343.73 MiB

tps: 403.83 ± 2.4315% (std=5.73)
qps: 403.83 ± 2.4315% (std=5.73)
lat_avg: 633.83 ± 2.3937% (std=8.95)
lat_99th: 1412.98 ± 2.5748% (std=17.34)
```

there is batch insert benchmark result, @zz-jason PTAL

**Comment by mahjonp** [comment]
Although we haven't found the reason why e954a5a caused a performance regression, but we can ensure that this PR will not impact sysbench benchmark result, so can we merge it first? @zz-jason

**Pull request merged**